staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/coerce.go

```diff
@@ -18,40 +18,78 @@ package objectmeta
 
 import (
 	"fmt"
+	"strings"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	utiljson "k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/klog/v2"
 )
 
 // GetObjectMeta does conversion of JSON to ObjectMeta. It first tries json.Unmarshal into a metav1.ObjectMeta
 // type. If that does not work and dropMalformedFields is true, it does field-by-field best-effort conversion
 // throwing away fields which lead to errors.
 func GetObjectMeta(obj map[string]interface{}, dropMalformedFields bool) (*metav1.ObjectMeta, bool, error) {
+	meta, found, _, err := GetObjectMetaStrict(obj, dropMalformedFields, false)
+	return meta, found, err
+}
+
+// GetObjectMetaStrict allows for strict unmarshaling of the metadata bytes, returning any unknown field paths
+// it encounters
+func GetObjectMetaStrict(obj map[string]interface{}, dropMalformedFields, strict bool) (*metav1.ObjectMeta, bool, []string, error) {
 	metadata, found := obj["metadata"]
 	if !found {
-		return nil, false, nil
+		return nil, false, nil, nil
 	}
 
 	// round-trip through JSON first, hoping that unmarshalling just works
 	objectMeta := &metav1.ObjectMeta{}
 	metadataBytes, err := utiljson.Marshal(metadata)
 	if err != nil {
-		return nil, false, err
+		return nil, false, nil, err
 	}
-	if err = utiljson.Unmarshal(metadataBytes, objectMeta); err == nil {
-		// if successful, return
-		return objectMeta, true, nil
+	if strict {
+		var strictErrs []error
+		strictErrs, err = utiljson.UnmarshalStrict(metadataBytes, objectMeta)
+		if err == nil {
+			if len(strictErrs) > 0 {
+				unknownPaths := make([]string, len(strictErrs))
+				for i, se := range strictErrs {
+					// each strict error should follow the format `unknown field "path.to.unknown.meta.field"`.
+					// Because we are isolating the metadata block of the original object, we have to add it back
+					// to the path here (because UnmarshalStrict has no way of knowing that the bytes it is unmarshaling
+					// are from the metadata)
+					splitErr := strings.Split(se.Error(), `"`)
+					if len(splitErr) == 3 && splitErr[0] == "unknown field " && splitErr[2] == "" {
+						unknownPaths[i] = "metadata." + splitErr[1]
+					} else {
+						// this should never be reached.
+						// If it does, there's an issue with how we're formatting
+						// strict errors and we should just return as if strict is false.
+						klog.Warningf("unrecognized strict err format %s, falling back to non-strict unmarshaling", se.Error())
+						return objectMeta, true, nil, nil
+					}
```

Comment:
hrmm... this translation looks pretty rough / fragile...

in the strict case, it seems like we could do this to get the actual fieldPath we wanted to reflect in errors:

```diff
-metadataBytes, err := utiljson.Marshal(metadata)
+wrappedMetadataBytes, err := utiljson.Marshal(map[string]interface{}{"metadata":metadata})
```

and then UnmarshalStrict into a locally defined type that only defines a single field:

```go
Metadata metav1.ObjectMeta `json:"metadata"`
```

would something along those lines get us the field paths we want in the strict errors for free?

Reply:
See discussion offline, but I think there’s a better way to do this that involves a few changes to kjson. See my [proof of concept](https://github.com/kevindelgado/kubernetes/pull/10) and the corresponding [kjson PR](https://github.com/kubernetes-sigs/json/pull/16)